controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,287 @@
+func (r *HostFirmwareComponentsReconciler) updateHostFirmwareComponents(components []metal3api.FirmwareComponentStatus, info *rhfcInfo) (err error) {
+	dirty := false
+	var newStatus metal3api.HostFirmwareComponentsStatus
+
+	// change the Components in Status
+	newStatus.Components = components
+	// Log the components we have
+	info.log.Info("firmware components for node", "components", components, "bmh", info.bmh.Name)
+
+	// Check if the updates in the Spec are different than Status
+	updatesMismatch := !reflect.DeepEqual(info.hfc.Status.Updates, info.hfc.Spec.Updates)
+	// Check if the components information we retrieved is different from the one in Status
+	componentsInfoMismatch := !reflect.DeepEqual(components, info.hfc.Status.Components)
+
+	reason := reasonSuccess
+	generation := info.hfc.GetGeneration()
+
+	if updatesMismatch {
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+
+		err := r.validateHostFirmwareComponents(info)
+		if err != nil {
+			info.publishEvent("ValidationFailed", fmt.Sprintf("Invalid Firmware Components: %s", err))
+			if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionFalse, reasonConfigurationError, "Invalid Firmware Components") {
+				dirty = true
+			}
+		}
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsValid, metav1.ConditionFalse, reason, "") {
+			dirty = true
+		}
+	}
+
+	if componentsInfoMismatch {
+		if setUpdatesCondition(generation, &newStatus, info, metal3api.HostFirmwareComponentsChangeDetected, metav1.ConditionTrue, reason, "") {
+			dirty = true
+		}
+	}
+
+	// Update Status if has changed
+	if dirty {
+		info.log.Info("Status for HostFirmwareComponents changed")
+		info.hfc.Status = *newStatus.DeepCopy()
+
+		t := metav1.Now()
+		info.hfc.Status.LastUpdated = &t
+		return r.Status().Update(info.ctx, info.hfc)
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareComponentsReconciler) SetupWithManager(mgr ctrl.Manager, maxConcurrentReconcile int) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3api.HostFirmwareComponents{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: maxConcurrentReconcile}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+			}).
+		Complete(r)
+}
+
+func (r *HostFirmwareComponentsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+	r.Log.Info("hostfirmwarecomponents in event handler")
+
+	if e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration() {
+		r.Log.Info("returning true as generation changed from event handler")
+	}
+
+	return false
+}
+
+func (r *HostFirmwareComponentsReconciler) validateHostFirmwareComponents(info *rhfcInfo) error {
+	allowedNames := map[string]struct{}{"bmc": {}, "bios": {}}
```

Comment:
We don't validate the values coming from Ironic, do we? So it's possible that a future driver or a change to Redfish adds more components. Do we really want to update BMO every time?